sklearn/cluster/kernel_kmeans.py

```diff
@@ -0,0 +1,279 @@
+"""
+    Kernel K-means clustering
+
+    Reference
+    ---------
+    Kernel k-means, Spectral Clustering and Normalized Cuts.
+    Inderjit S. Dhillon, Yuqiang Guan, Brian Kulis.
+    KDD 2004.
+
+    Author: Mathieu Blondel <mathieu@mblondel.org>
+            Ishank Gulati <gulati.ishank@gmail.com>
+
+    License: BSD 3 clause
+"""
+from __future__ import print_function
+from __future__ import division
+
+import numpy as np
+
+from ..base import BaseEstimator, ClusterMixin
+from ..metrics.pairwise import pairwise_kernels
+from ..utils import check_random_state
+from ..externals.six.moves import xrange
+from ..utils import check_array
+from ..utils.validation import FLOAT_DTYPES
+
+
+class KernelKMeans(BaseEstimator, ClusterMixin):
+    """
+    Parameters
+    ----------
+
+    n_clusters : int, optional (default=3)
+        The number of clusters to be formed as well as number of centroids
+        to be generated.
+
+    max_iter : int, optional (default=300)
+        Maximum number of iterations of algorithm for a single run.
+
+    n_init : int, optional (default=10)
+        Number of random initializations.
+
+    tol : float, optional (default=1e-3)
+        Tolerance for stopping criterion.
+
+    kernel : string, optional (default='linear')
+        Specifies the kernel type to be used in the algorithm.
+        It must be one of 'linear', 'poly', 'rbf', 'sigmoid', 'precomputed' or
+        a callable.
+        If none is given, 'linear' will be used. If a callable is given it is
+        used to pre-compute the kernel matrix from data matrices; that matrix
+        should be an array of shape ``(n_samples, n_samples)``.
+
+    gamma : float, optional (default='auto')
+        Kernel coefficient for 'rbf', 'poly' and 'sigmoid'.
+        If gamma is 'auto' then 1/n_features will be used instead.
+
+    degree : int, optional (default=3)
+        Degree of the polynomial kernel function ('poly').
+        Ignored by all other kernels.
+
+    coef0 : float, optional (default=1.0)
+        Independent term in kernel function.
+        It is only significant in 'poly' and 'sigmoid'.
+
+    kernel_params : optional
```

Comment:
should be ``dict, optional``

Also, we should mention here that this is only referenced if the ``kernel`` is a callable.